extensions/amp-viewer-integration/0.1/messaging/messaging.js

```diff
@@ -116,15 +116,85 @@ export class WindowPortEmulator {
  * ampdoc and Bob is the viewer.
  */
 export class Messaging {
+  /**
+   * Performs a handshake and initializes messaging.
+   * @param {!Window} source
+   * @param {!Window} target
+   * @param {?string=} opt_token
+   * @param {?number=} opt_interval
+   * @return {!Promise<!Messaging>}
+   */
+  static initiateHandshakeWithDocument(
+    source,
+    target,
+    opt_token,
+    opt_interval
+  ) {
+    const interval = opt_interval || 1000;
+    return new Promise(resolve => {
+      const intervalRef = setInterval(() => {
+        const channel = new MessageChannel();
+        const pollMessage = /** @type {JsonObject} */ ({
+          app: APP,
+          name: HANDSHAKE_POLL_MSG,
+        });
+        target./*OK*/ postMessage(pollMessage, '*', [channel.port2]);
+
+        const port = channel.port1;
+        const listener = e => {
+          const data = {e};
+          if (data['app'] === APP && data['name'] === CHANNEL_OPEN_MSG) {
+            clearInterval(intervalRef);
+            port.removeEventListener('message', listener);
+            const messaging = new Messaging(null, port, false, opt_token, true);
+            messaging.sendResponse_(data['requestid'], CHANNEL_OPEN_MSG, null);
+            resolve(messaging);
+          }
+        };
+        port.addEventListener('message', listener);
+        port.start();
+      }, interval);
+    });
+  }
+
+  /**
+   * Waits for handshake from iframe and initializes messaging.
+   * @param {!Window} source
+   * @param {!Window} target
+   * @param {string} origin
+   * @param {?string=} opt_token
+   * @return {!Promise<!Messaging>}
+   */
+  static waitForHandshakeFromDocument(source, target, origin, opt_token) {
+    return new Promise(resolve => {
+      const listener = e => {
+        if (
+          e.origin == origin &&
+          (!e.source || e.source == target) &&
+          e.data['app'] === APP &&
+          e.data['name'] === CHANNEL_OPEN_MSG
+        ) {
+          source.removeEventListener('message', listener);
+          const port = new WindowPortEmulator(source, origin, target);
+          const messaging = new Messaging(null, port, false, opt_token, true);
+          messaging.sendResponse_(e.data['requestid'], CHANNEL_OPEN_MSG, null);
+          resolve(messaging);
+        }
+      };
+      source.addEventListener('message', listener);
+    });
+  }
+
   /**
    * Conversation (messaging protocol) between me and Bob.
-   * @param {!Window} win
+   * @param {?Window} win
```

Comment:
Hmm. Really not a huge fan of this, to be honest. Is this really the only use for the `win` property in this class?

Reply:
It seems so. Doing it this way also allowed me to remove that parameter from `initiateHandshakeWithDocument`.